### ISSUE TITLE
[Bug] Getter needlessly recomputing

### ISSUE THREAD

**Issue by boris-petrov** [emberjs/ember.js#20050]
### 🐞 Describe the Bug
See `Minimal Reproduction`.

### 🔬 Minimal Reproduction
[Here is a repo](https://github.com/boris-petrov/ember-recompute-issue). Clone, `yarn`, `ember serve`. Clicking on the link logs `computing values` on every click. It shouldn't because the getter doesn't touch any of the values that are changed on click. It is even `@cached` (though the same thing happens without that too).

Removing `@tracked` on `bool` fixes the issue... which doesn't make much sense because that variable is not touched. Not passing it as an argument to the component also fixes it.

### 😕 Actual Behavior
Getter recomputes.

### 🤔 Expected Behavior
Getter to not recompute.

### 🌍 Environment

- Ember: 4.3.0
- Node.js/npm: 17.8.0
- OS: Linux
- Browser: Chrome

**Comment by gnclmorais** [comment]
I ran your project locally (thank you for providing one!) and I noticed that by simply changing that `bool` variable name to something else (like `stuff`), the recomputation doesn’t happen anymore.

Also, if when the component loads, you add `{{this.bool}}` to your template file, you’ll see it outputting `true` straight away.

Is there perhaps a `bool` variable on Glimmer components from the get-go? 🤔

**Comment by bertdeblock** [comment]
> I noticed that by simply changing that bool variable name to something else (like stuff), the recomputation doesn’t happen anymore.

Did you also rename it on the invocation side?

```hbs
{{foo-component array=this.array bool=true}}
```

Maybe when you mutate the `array` arg (naughty) the `bool` arg is "dirtied" as well somehow?
Seems to work as expected when using a Glimmer component though.

**Comment by chancancode** [comment]
This is an.. interesting.. combination. I think it’s actually working as intended, just an unexpected combination. In classic components, when any of the args changes the component manage does `setProperties` on all the args. Usually this goes through the legacy `.set` which does an `===` to avoid invalidating computed properties. But since you made it `@tracked` it’s now a setter and bypasses (the rather complicated) legacy `.set` logic.

so TL;DR it’s all “working as intended” since you did actually update the tracked property, just not visibly through the component manage. I think the general advise (other than refactor away from the classic component) is don’t mark classic component arguments as tracked?

what were the cold trying to achieve in the real use case?

**Comment by boris-petrov** [comment]
@gnclmorais - I believe @bertdeblock is correct - you've probably forgotten to rename the variable in `application.hbs`.

@chancancode - thanks for the detailed explanation. What you say makes sense and yes, in that sense this is working as expected. Definitely not perfect though. The reason I wanted `bool` to be tracked is because it's used in other getters and if it changes, for them to recompute, it has to either be `@tracked` or the getters should be `@computed`. I just thought that it's better to use `@tracked`. Never did I imagine that changing one argument would cause a modification to be done on all of them. :) Not sure if this is something that could be fixed in Ember or even if anyone would want to invest time in it.

Thanks for the help! I guess the issue could be closed but I'll leave it open if someone wants to say something more or attempt a fix or something.

**Comment by chancancode** [comment]
Yeah, I think it "makes the kind of sense it does", when mixing the new and old world. It would be one thing it's an error, but I think this is just an unfortunate interaction that falls out for all the semantics. Specially, you are in the middle of a migration in this case – if you are using all classic idioms, it works, if you are using all octane idioms, it also works, but if you are mixing them them you need to take special care, hence things like `@dependentKeyCompat` and such.

Anyway, for your specific case, I believe you can actually

**Comment by chancancode** [comment]
I'll close the issue to indicate the current thinking/status, but feel free to continue the discussion and if I am wrong about something happy to re-open as well.

**Comment by chancancode** [comment]
> Never did I imagine that changing one argument would cause a modification to be done on all of them. :)

To be clear, this is generally not the mental model. [This](https://github.com/emberjs/ember.js/blob/a0be65cc788ce61576b08107454ec7beb22baa30/packages/%40ember/-internals/glimmer/lib/component-managers/curly.ts#L442) is where the `setProperties` is, which loops through all the arguments and call `.set()` on each of them. In all-classic world, this is not a big deal, since `.set()` no-ops when the value is the same. But if you turn a property into a getter/setter pair (which is what `@tracked` does) then I believe the setter does get called consistently (I didn't actually verify this), which is also a generally desirable behavior I think.

**Comment by boris-petrov** [comment]
> at some point it's probably just better and less confusing to port the component to use `@glimmer/component` instead of leaving it in an in-between state with caveats

For sure, it's just legacy code which is definitely not easy to port.

In any case, thanks a lot for the discussion! Your suggestions above gave me the idea to just use [dedupeTracked](https://github.com/tracked-tools/tracked-toolbox#dedupetracked). Interestingly though, it doesn't work, which is very strange but I don't have the time to debug it now. Thanks again for the help!
